### ISSUE TITLE
Where did the source go?

### ISSUE THREAD

**Issue by haarp** [afollestad/cabinet-issue-tracker#102]
What happened to the source code? Is this app closed-source now? Including the contributions you received from others?

Just curious!

**Comment by joho506** [comment]
It is indeed closed-source now, IIRC someone copied the entire code including crash reporting and didn't bother to change the account it registers with, so he kept getting constant emails every time the rip-off (com.grizzly.main) was opened.

**Comment by haarp** [comment]
That's a license issue tho, not a source issue. And naturally it's possible that some idiots copy everything, but in the end that got taken care of, didn't it?

Unfortunately I can't update this app anymore now. Which is sad, I liked it.

Closing the source is also a bit rude, considering that others also have contributed to this app when it was still open.

**Comment by afollestad** [comment]
Actually I wasn't getting enough contribution for it to make a difference.
The 2 people who made frequent contributions still have access to the
source.

**Comment by haarp** [comment]
Oh, I see, nice.

Do you think you might open it up again at some point?

**Comment by afollestad** [comment]
Probably not, I may have paid features at some point so it's safer to keep
it as my intellectual property.

**Comment by haarp** [comment]
I see. Just a note: Open-source doesn't imply that it must cease being your IP. It all depends on the license used.

**Comment by afollestad** [comment]
Well everyone being able to compile it and disable having to pay for things
in the future would defeat the purpose

**Comment by haarp** [comment]
Well to be fair, there aren't many people able to do that. There are some apps (OsmAnd comes to mind) that officially offer a full open-source variant for the 0.1% users using F-Droid and a paid version for the other 99.9% in the Play Store :)

Alternatively, paid features could also reside on a private development tree with the rest being public.

But it's your choice of course.

**Comment by afollestad** [comment]
Hmm, yeah that's a good point

**Comment by th3m4ri0** [comment]
Agree with haarp. I saw some applications in the past do the open source free / closed source premium. But again it's your choice. Keep up the good work :octocat:

**Comment by geecko86** [comment]
You licensed the program under the [X11 license](http://directory.fsf.org/wiki/License:X11) also known as the MIT license. Out of curiosity, don't you *have* to leave outdated versions of the code public?

I'd rather use that.

**Comment by afollestad** [comment]
I don't think anything says I have to put my property out in the public if it previously was open source.

**Comment by geecko86** [comment]
> Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software

Well, I'd like to use this permission. This is how it was licensed to me and you can't just go back on your word.

**Comment by plusCubed** [comment]
@geecko86 The key there is that you have the permission to use the a copy of the software as described, but it doesn't say that you have the right to obtain a copy of the software. If you have a copy of the source with the license, you may use it as said in the license. As long as Aidan has copyright on the software, he can change the license and is not required to provide a previous copy of the source. 

That said, there are forks of the previously open source code for Cabinet on Github that you can get.

**Comment by afollestad** [comment]
Most of which are out dated, or I already requested a takedown for.

**Comment by nyuszika7h** [comment]
@afollestad What do you mean requested a takedown? You clearly don't understand how open source licenses work. Since you seem to be the sole copyright holder, you can choose to relicense future versions as you wish, but open source licenses like the MIT license **cannot be revoked**, meaning everyone is free to continue redistributing and modifying the versions that had their source code published, as long as they preserve the attribution.

**Comment by nyuszika7h** [comment]
@afollestad What do you mean requested a takedown? Open source licenses like the MIT license **cannot be revoked**. If you're the sole copyright holder, you can choose to grant any license to anyone you wish, or even stop licensing future versions and keep the source code your own IP. But you cannot revoke previously granted open source licenses, as none of them allow revocation, for a very good reason.

**Comment by geecko86** [comment]
Agreed. I understand that it's hard to swallow but it is your responsibility to keep your word. You really cannot do this.

**Comment by afollestad** [comment]
I'm aware. If you find the old source, go ahead. Like I said, I don't keep copies of old source code around.

**Comment by plusCubed** [comment]
Whoever has a fork is free to continue using it or distributing it - like you guys said, the license grants them the power to do so. Aidan is just requesting/asking them to take it down, not using legal action (I think takedown is the wrong word. Nobody has the legal rights to do so, like you guys said).

**Comment by Techman-** [comment]
What a betrayal

**Comment by afollestad** [comment]
Access to source for an app I've spent countless hours on for 2-3 years as
a high school and college student is a privilege, not a right.

**Comment by Techman-** [comment]
I used to support this app, and even recommend (push) this to my friends. Then I hear that you just pull the source code out of nowhere, which honestly is one of the biggest douchbag moves a developer can make. Why are you making an asshole out of yourself?

**Comment by afollestad** [comment]
I'm not. People were stealing my code and like being able to claim
something I've put a lot of work into. I haven't gotten any benefits from
when it was open source at all. Its future is better closed source.

**Comment by afollestad** [comment]
Sorry, and I like*

**Comment by Techman-** [comment]
No, open source worked fine. That's not the issue. There is a seriously bad problem with your thought process. So instead of doing the proper thing about folks who are trying to parade around as you, you decide to show that you're the weaker man by pulling the source code? Man up and deal with problems the proper way. This also just shows me and everyone else how much of a rookie developer you are.

Being open source was a feature, honestly. Cabinet was a pretty nice material file manager, **for something that was open source**. But hey, since you're now proprietary, I guess I can just choose an [obviously better choice](https://play.google.com/store/apps/details?id=nextapp.fx). FX is proprietary too, but they have a decent design and a remote networking features **that actually work**.

So no, going closed didn't help you or anyone. It was just a ~~bitch~~ weak move.

**Comment by haarp** [comment]
Please remain civil. Yeah, someone might not quite understand the point of open-sourcing their software, and may have made some poor decisions, but that doesn't mean we have to be abusive.

Aidan, what did you actually expect would happen? The ability to "steal" code is one of the main principles of open source. The situation you were in can happen, but unless the person forking your code does a better job than you, they won't succeed. The fact that you received bug reports for the fork would've been easily fixed by simply not keeping your API key in the repository. And if you really, really want to go against the spirit of free software, you can use a license that prohibits using your code in other projects. As I said, it was a license problem, not an open-source problem.

Seeing how you feel about the whole thing, why did you use open-source in the first place?

**Comment by afollestad** [comment]
I had made it open source because I was starting my first year of college
and thought the project would get contributions and benefit, which was not
the case.

Now that I'm getting job offers at Facebook and other big companies, and
want to intern at Google, I decided to reclaim my property since nothing
good came from open source. Daniel was the only main contributor so he
became a dev partner and still has access among a few others.

**Comment by afollestad** [comment]
I stand by decisions. The only douche bag here is the ones who obviously
don't understand what it's like the be the dev of a very popular project
that has a lot of competitors. Boid for Twitter was similar when I was
working on that.

**Comment by nyuszika7h** [comment]
Not trying to convince you now, but for the record, using a license the GPL mitigates the "stealing your code" somewhat. Permissive licenses like the MIT license require attribution, but they allow other people to legally redistribute a proprietary version of your app, as long as it has attribution somewhere.

**Comment by afollestad** [comment]
Like I said above, the code when that was the license is free to use.

**Comment by nyuszika7h** [comment]
How is that relevant to my comment?

**Comment by afollestad** [comment]
The code after the point where it was removed is not. That's how it works.

**Comment by nyuszika7h** [comment]
I have a feeling we're talking about completely different things here. I was just saying that if you ever consider open source again (not necessarily for this project, can be any other one too), you may like the GPL better than permissive licenses.

**Comment by afollestad** [comment]
Sorry, right now I want to keep it private. Theft was bad to the point
where someone forgot to remove their copied version from the crash
reporting system so I got an email saying an app with a different name,
package name, and icon was registered. Competitors were also beginning to
copy a few of my components.

**Comment by afollestad** [comment]
People are still gonna steal regardless of the license.

**Comment by nyuszika7h** [comment]
If they "steal" in a way that the license doesn't allow, there's always DMCA takedown.

**Comment by afollestad** [comment]
Btw @Techman-, I'm almost 20 and I've been programming for almost 8 years. I'm self taught, employed as a software dev, and I've taken 5 CS classes (well skipped past 3, and taken 2). Watch who you're calling a noob, dumbass.

**Comment by geecko86** [comment]
> People were stealing my code

> Theft

I'm sure you can see how saying that about public code is not very coherent..

> Now that I'm getting job offers at Facebook and other big companies, and
want to intern at Google

That's very relevant. We don't care.

> People are still gonna steal regardless of the license.

So you're actually saying changing the license was pointless?

Now, some of us *"thieves"* (free software & transparency advocates would be more like it) are willing to fork the code. Could you please inform us regarding *where* we can find the code, what the license for it is and what we would be required to change? For instance, I suppose we're not allowed to use the name "Cabinet" or use your icon.

Obviously, I completely respect your decision to make the code proprietary, that's up to you and good for you if you get something out of that. What I don't understand is why you'd deceive people into taking down the source code that you made public.

**Comment by afollestad** [comment]
Your attitude doesn't make me want to respond to you at all. It is relevant, it's one of a few reasons I decided to take my property back. It was my decision in the first place to start this project, I'm the one that spent 2 years slowly working on it, not any of you. 

Find it yourself, I'm done arguing. Search GitHub for Cabinet.

**Comment by afollestad** [comment]
The second result is an old repository that isn't up to date, but it's something.

https://github.com/search?utf8=%E2%9C%93&q=cabinet

**Comment by geecko86** [comment]
Alright, that's something. :)

It doesn't say under which license it was released. That's obviously the core issue. Could you answer that question? This is version 1.7.7, can I ask you what was the latest open source version of Cabinet?

After that we can say goodbye. Sad to see that discussing this matter bothered you so much.

PS: Security wise, [this](https://puu.sh/h313O.png) isn't very good practice.. I hope you don't use that password anymore.

**Comment by plusCubed** [comment]
@geecko86 The point of open sourcing Cabinet was so people can contribute and make the software better, e.g. what you are saying with transparency. Creating "new" versions that are completely based on someone's code isn't. I'm sure there's a license out there that specify this in some form or another, but without time spent on legal issues and takedowns, it's basically the same as no license. 

I see a few general reasons someone might want the source of Cabinet: 
1. Make improvements. 
2. See the code and know that it's not malware/know it's implemented well
3. Use some part of the code in their own projects
4. Use all of the code and start a "new" file manager

By making Cabinet proprietary, these things are no longer possible. I'll break down the pros and cons here:
1. Neutral, because someone like me (before the transition to closed source) couldn't have made the changes I did. However, point is, not a lot of people were doing this.
2. Con. People can no longer verify it's safe or implemented well and have to trust the developer. I don't have a refute for this. 
3. Neutral. People can no longer just use a useful part of the code. However, I'm certain if you just ask Aidan or I we will be happy to explain how something is done. As always, knowledge should be spread, not contained - by making Cabinet open source, it just makes it a bit harder to obtain that knowledge, but people who are actually interested can.
4. Pro. People can't just copy the code, and build something on top of it instead of making improvements to the original code.

TL;DR: If you would like to make improvements, you can still do so through submitting feature requests, and if you are really committed, it's Aidan's decision to let you in. If you want to know how a part is implemented, we can explain (and even show you the relevant code). If you want to copy and create a "new" version, GTFO (excuse the obscenity).

If I missed any reason or if there are any fallacies in this post, please point them out.

**Comment by plusCubed** [comment]
@geecko86 Actually, that post is directed towards everyone, not you. Sorry if that was misleading. I've removed the @ tag.

**Comment by afollestad** [comment]
@geecko86 sorry I actually can't remember what license it was, I believe it was MIT though

**Comment by geecko86** [comment]
@plusCubed I am not arguing against making it close source. He can obviously make this decision and I don't tell people what to do. The only thing I'm here for is information regarding how I can fork this in the best possible way, not to tell people what to do.

I will also correct people when they make stuff up to fit their goals. Contrary to what you just said, copying and forking code is a completely normal thing to do in the world of free software. In fact, it's pretty much all we do. If you hate that idea, don't make it open source. And it's obviously not the decision he made. This is not the debate and I'm afraid you are wrong regarding what the consequences of making OSS are. Examples of open source projects being used without the agreement of their original creators are countless (Cyanogen, Blink, Kindle, ...) and it doesn't matter at all.

@afollestad I also believe it was MIT. Great, thanks!

Now that I know this, I will act in consequence. Meaning I'll include the copyright info with your code, as required per the MIT license. See, no stealing? ;)

> Copyright Aidan Follestad © 2014

I really don't see the point of using that language either, especially if you're going to claim things like "don't copy OSS or GTFO".

**Comment by afollestad** [comment]
By stealing, I mean copying the code and republishing it under different
app names without any attribution AND after the license was taken away.

**Comment by Techman-** [comment]
This developer is so ignorant it is hilarious. I don't care how many years you claim was spent programming. What mattera is that you're still a rookie developer because you make rookie decisions. Your stats don't phase me. Nobody cares what you did (allegedly) since 12 years of age.

In my opinion, snatching back code that was previously open should be illegal. But the choice of license is what controls that outcome. Face it: you made a weak decision and you must own up to it, unless you want to lose credibility. 

I don't like how you throw around Cabinet as your property.it might have been your project but you make it seem like you're using others' contributions just to improve something that is still yours.

Instead of using the law to deal with people who straight up copy you, you pulled the code. You just happened to be the next ~~clown~~ developer to make a mistake.

**Comment by afollestad** [comment]
@Techman- allegedly? It's fact. I'm not ignorant, I know the facts and these are decisions I spent weeks thinking about. 

I am owning up to it, all the comments above describe the reasoning behind it. It is not your decision, fucktard (sorry but if you're gonna try to put people down then I don't give a fuck about obscenity). 

If you would've come in here with a better attitude in the first place, maybe I would've changed my mind.

**Comment by plusCubed** [comment]
@geecko86 Thank you for pointing that out. Yes, I reflect what Aidan said - by copying I meant republishing. I agree that OSS should be allowed to be copied and fork - case in point, my own projects are open source, and I understand the consequences. In this particular case, Cabinet, Aidan's point of open sourcing didn't match up with the concept of open sourcing, and so it was changed - I believe we both agree. I think for that last comment, I meant to say "don't republish without attribution and changes or get out" - and yes, I don't like and rarely ever use obscenities. Thank you voicing your opinions.

**Comment by Techman-** [comment]
>fucktard

Pretty, pretty. About time you let your emotions flow through you. You wouldn't have changed your mind whether my approach was positive or not. Changing your decision now would make you look insecure (as if you wasn't already).

Call me what you want, but I haven't (nor would I) make rookie mistakes like these. Based on your comments from earlier, this looks like an action that didn't take weeks to think about. That's what you claim, right? Spending weeks to hit a delete button, knowing that it was going to cause a flare up?

You might see me as a dick, an asshole, or whatever you want. But you better bet that there's plenty out there that think like me and express the same feelings as me on this subject, but have decided to take a seat on this. Don't think I'm the only one, for that would be a foolish decision.

**Comment by afollestad** [comment]
@Techman- like you would know. Being negative isn't the way to make friends or try to sway people into doing what you want.

They're not rookie mistakes, they're decisions of an experienced software developer who changed his mind because he thinks he made a mistake making it open source in the first place. Do you really want to see my code that badly? Make your own damn app, do you not know enough to make one yourself or what?

And you come across as one, this wouldn't have turned into an "argument" if you didn't make it one. I honestly would give kind people access to my code if they requested it and had a good reason to use it.

**Comment by plusCubed** [comment]
@ Everyone: In summary, if you want to use the code or any part of it, or if you want to see it's not malware, or if you want to know how some part is implemented, or if you want to make contributions, or anything else that's not outright robbery, please just ask through message or email with a good reason. We are not trying and we don't want to lock down the source, we are only trying to prevent people from ignoring licenses and republishing.

**Comment by Techman-** [comment]
>Being negative isn't the way to make friends or try to sway people into doing what you want.

Very true. My intention wasn't to make friends or sway you back to open source. You think open source is pointless, so stick to your thoughts. I'm sure the open source community won't be missing you very much.

If I developed something for Android, it sure as hell wouldn't be a file manager. FX is already good and their price is reasonable. There are countless others on the Play Store that do the job as well. The IRC app scene is kinda dead on Android; Why don't you try something there?

**Comment by afollestad** [comment]
Obviously not. 

And I'm aware. This was an experiment that I started for Android Dev practice two years ago. I have other apps, some of which are under private testing, others which are done for personal clients, and others that are done for my employer. I don't really care what you think about the "file manager scene." People like Cabinet, so I will continue ti support it.

**Comment by geecko86** [comment]
@plusCubed Wouldn't it make more sense to use the proper means (aka takedown requests) to take care of that, though? Yes it would. And even if some manage to use your MIT code without citing your name, is it really that much trouble?

I agree that there's no use being so agressive, @Techman- . I agree that pride and self-righteousness have a part in this discussion, but it's not our place to tell him how to behave. Your comments haven't been very relevant either.

**Comment by afollestad** [comment]
Now, unless you have something nice to say, go away. You're not making anything better here.

**Comment by afollestad** [comment]
@Techman- also, why are you here in the first place if you wouldn't even develop a file manager? There's literally nothing you could use the source for. 

Anyways, if you have nothing nice to say, go away. You're completely irrelevant if you're just gonna rip on people.

**Comment by Techman-** [comment]
You want something nice? I've already said something nice about the app. I used to like it before I heard that you pulled this stunt.

I'd be interested in seeing what you could pull off as an IRC app, but then again you might pull another "reclaiming my property" move on everyone, if you decide to open source again. It's really not that hard to have open source software. Next time, pick a license that is a little more restrictive for forks. When a project takes your code and copies it flat-out, send a takedown request in and the other software is history (as far as `x` location is concerned).

I would love to see [IRCCloud for Android](https://github.com/irccloud/android), reimagined as an entirely-material app.

**Comment by geecko86** [comment]
A lot of people who don't code appreciate free software because of its nature. He's not complaining because he wants to contribute to the project, but rather because he is someone who doesn't like proprietary software.

Alright @Techman- , this is really out of topic now. You've made your point, he picked the wrong license and he already admitted that was a mistake.

**Comment by afollestad** [comment]
I understand that, but I don't understand what led him here and made him
decide to start a very long and pointless argument. His point would've been
more effective if he didn't come across as an asshole.

Anyways, it's late here and I have to get up for college classes so I'm off
to bed. I'll respond to anything else tomorrow.

**Comment by afollestad** [comment]
One more thing, I'm also a fan of open source software. My servers use
Linux distributions, and I used desktop Linux distros for quite a while.
There's just some cases where proprietary software is better for its own
health in the community of apps.

**Comment by Techman-** [comment]
@geecko86 Yeah the IRC stuff is off topic. Too bad @afollestad doesn't have an IRC channel (Maybe I  might be able to help with that?)

**Comment by geecko86** [comment]
Problem is you haven't been very convincing in your explanation as to why it was better for your file manager to be close sourced.. Hence the genuine disappointment and the backlash.

**Comment by plusCubed** [comment]
@geecko86 See the [Material Manager fiasco](https://plus.google.com/+AidanFollestad/posts/YeZ9Kasocnp) a few months back. It's just the complications with takedowns - writing them and constantly looking for republished software on the internet instead of working on improvements. There's also the problem of not being able to identify the republished ones (e.g. foreign language, different names, icons etc). In the end it's easier to maintain as proprietary software - granting access only to those who aren't going to misuse it. For those who have the time and resources to maintain their open source software (e.g. my own apps that aren't super popular and aren't going to be republished without being found), I fully support and see the appeal in that.

**Comment by geecko86** [comment]
The G+ post is not public.. :/ 
So what was going on? People were removing the MIT disclaimer from Aidan's .java files on purpose and claiming it as their own? That's... disturbing.

**Comment by plusCubed** [comment]
@geecko86 Oh, join the Cabinet Google+ community! So basically someone republished the app with some improvements but charged for some open source features of Cabinet (e.g. root access). The MIT license required the person using to distribute a copy of the license and a copyright notice, and after arguing a bit the developer added credit in the app, but later removed it because he claimed he had enough of rewritten code that it couldn't be credited towards Aidan. Since his software's closed source though, we can't actually see how much he rewrote (there's no evidence because from a GUI standpoint it looked the same). By that point, since Aidan doesn't have enough evidence, Aidan just stopped trying to chase him down.

**Comment by Techman-** [comment]
You should have not even tried negotiating with the developer. You should have sent a takedown request before the other dev had a chance to get away with it.

**Comment by plusCubed** [comment]
@Techman- I didn't even know about Cabinet at that point yet lol. And yes, I believe Aidan and many users of Cabinet sent takedowns to Google but they didn't reply. Again, I wasn't involved yet at that point - Aidan can clarify later.

**Comment by geecko86** [comment]
I'll have a look when I get the time. Surely it's easy enough to decompile.

**Comment by afollestad** [comment]
@geecko86 I wouldn't do that with the latest version. It's obfuscated. I will also have the right to legal action if you do. I have no problems with the MIT licensed code though.

**Comment by afollestad** [comment]
Disregard, just woke up and didn't read everything yet.
